python2_nbd_client.py

```diff
@@ -52,17 +57,58 @@ def close(self):
             self._disconnect()
             self._closed = True
 
-    def _non_fixed_new_style_handshake(self, export_name):
+    def _upgrade_socket_to_TLS(self):
+        # Default context uses PROTOCOL_TLS
+        thesocket = self._s
+        context = ssl.create_default_context()
+        context = ssl.SSLContext(ssl.PROTOCOL_TLS)
```

Comment:
Can we use
ssl.PROTOCOL_TLSv1_2
here instead of leaving it up  to negotiation?

Reply:
Sure

Reply:
That option is deprecated: the documentation recommends excluding each pre-1.2 version rather than mandating 1.2 (so I suppose this way a future version will be allowed automatically).
https://docs.python.org/3/library/ssl.html#constants

> ssl.PROTOCOL_TLSv1_2
> 
> Selects TLS version 1.2 as the channel encryption protocol. This is the most modern version, and probably the best choice for maximum protection, if both sides can speak it. Available only with openssl version 1.0.1+.
> 
> New in version 3.4.
> 
> Deprecated since version 3.6: OpenSSL has deprecated all version specific protocols. Use the default protocol PROTOCOL_TLS with flags like OP_NO_SSLv3 instead.

Reply:
Let's restrict to supported and secure ones using OP_NO_SSL... then.

Reply:
Using create_default_context already sets the OP_NO_SSL options:
`The settings are: PROTOCOL_SSLv23, OP_NO_SSLv2, and OP_NO_SSLv3 with high encryption cipher suites without RC4 and without unauthenticated cipher suites. `

Reply:
Please specify `OP_NO_TLSv1_1` and `OP_NO_TLSv1` as well.
See https://docs.python.org/2/library/ssl.html#ssl.OP_NO_TLSv1
(Yes, I know those options say they are "only applicable in conjunction with `PROTOCOL_SSLv23`" but `PROTOCOL_SSLv23` is an alias for `PROTOCOL_TLS`.)

Reply:
If I specify `context.options = ssl.OP_NO_TLSv1_1` then the connection fails with `ssl.SSLError: [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version` 

`context.options = ssl.OP_NO_TLSv1` is fine

Reply:
Which seems backwards given the error message

Reply:
That's very odd, given that (I believe) the server is accepting v1.2 only.

Reply:
I've tested a couple times and this is definitely the behaviour I'm seeing.

Reply:
Did it work (connect successfully) when you specified `PROTOCOL_TLSv1_2`?

Reply:
Yep that works.

Reply:
looks like there was also an issue in how I was specifying multiple NO_OP_* options so I was just overwriting them. FIxed now. Still working with `PROTOCOL_TLSv1_2`

Reply:
@robertbreker @thomassa 
I have now updated it to be in the form:
`        context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+        context.options &= ~ssl.OP_NO_TLSv1
+        context.options &= ~ssl.OP_NO_TLSv1_1
+        context.options &= ~ssl.OP_NO_SSLv2
+        context.options &= ~ssl.OP_NO_SSLv3`

Is that ok with you both? If I try specifying PROTOCOL_TLS I get errors

Reply:
That should work fine.
Specifying `PROTOCOL_TLSv1_2` means the context will use v1.2 only, so the I think the `OP_NO...` options are redundant (but harmless).
I suppose eventually the deprecated `PROTOCOL_TLSv1_2` will be removed and then this will stop working, but perhaps by then it will work with `PROTOCOL_TLS` together with the `OP_NO_...` options.

Reply:
ok I played around with this a bit more and it now works with PROTOCOL_TLS. I think the failurs I was seeing must have been due to an error in my syntax for adding multiple options which is now resolved. I think everything is fixed now